### PR TITLE
Properly search for users when limittogroups is enabled

### PR DESCRIPTION
**Pull request by rullzer**
Searching just for the uid is not enough.
This makes sure this done properly again now.

Actually fixes a valid failure in the integrationtests...

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]

**Comment by rullzer** [comment]
To test

* limit users to share only in their own groups
* create a group TestGroup
* create a user TestUser that belongs to TestGroup
* create a user TestUser2 that belongs as well to TestGroup

Now as TestUser search the sharee endpoint for 'test'

Before:
* see the group but not the user

After:
* see both

File: lib/private/Collaboration/Collaborators/UserPlugin.php [hunk 1]
```diff
@@ -77,13 +77,24 @@ public function search($search, $limit, $offset, ISearchResult $searchResult) {
...
+			$userGroups = $this->groupManager->getUserGroupIds($this->userSession->getUser());
+
+			$usersTmp = $this->userManager->searchDisplayName($search, $limit, $offset);
+			foreach ($usersTmp as $user) {
+				if (!$user->isEnabled()) {
+					// Ignore disabled users
+					continue;
+				}
+
+				$commonGroups = $this->groupManager->getUserGroupIds($user);
```
Comment (nickvergessen) [inline lib/private/Collaboration/Collaborators/UserPlugin.php:89]: Are you missing an intersect with `$userGroups` ?

Comment (rullzer) [reply lib/private/Collaboration/Collaborators/UserPlugin.php:89]: yes it is

Comment (nickvergessen) [reply lib/private/Collaboration/Collaborators/UserPlugin.php]: Fixed

**Comment by nickvergessen** [comment]
I can not reproduce the failure before the patch so unwilling to approve the fix, but at least it is not breaking something.

**Comment by rullzer** [comment]
Ok let me try to explain. Basically it does boom because of:

https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/Group/Group.php#L310

Which goes to: https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/Group/Database.php#L333

Now there are 2 issues here

1. It doesn't search in displaynames (despite the method names)
2. it does a regular like. Which is case sensitive.

The new way will actually search displaynames as well. Which is what you want.

**Review by skjnldsv — APPROVED** [review]
Smarter!

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
```
There were 12 errors:
138	
139	1) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #2 ('test', true, true, array(), array(), array(), array(), true, false)
140	Invalid argument supplied for foreach()
141	
142	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
143	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
144	
145	2) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #3 ('test', true, false, array(), array(), array(), array(), true, false)
146	Invalid argument supplied for foreach()
147	
148	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
149	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
150	
151	3) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #6 ('test', true, true, array(), array(), array(), array(), true, Mock_IUser_6288abb4 Object (...))
152	Invalid argument supplied for foreach()
153	
154	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
155	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
156	
157	4) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #7 ('test', true, false, array(), array(), array(), array(), true, Mock_IUser_6288abb4 Object (...))
158	Invalid argument supplied for foreach()
159	
160	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
161	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
162	
163	5) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #8 ('test', true, true, array('test-group'), array(array('test-group', 'test', 2, 0, array())), array(array('Test', array(0, 'test'))), array(), true, Mock_IUser_6288abb4 Object (...))
164	Invalid argument supplied for foreach()
165	
166	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
167	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
168	
169	6) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #9 ('test', true, false, array('test-group'), array(array('test-group', 'test', 2, 0, array())), array(array('Test', array(0, 'test'))), array(), true, Mock_IUser_6288abb4 Object (...))
170	Invalid argument supplied for foreach()
171	
172	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
173	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
174	
175	7) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #16 ('test', true, true, array('abc', 'xyz'), array(array('abc', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...))), array('xyz', 'test', 2, 0, array())), array(), array(array('Test One', array(0, 'test1'))), true, false)
176	Invalid argument supplied for foreach()
177	
178	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
179	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
180	
181	8) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #17 ('test', true, false, array('abc', 'xyz'), array(array('abc', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...))), array('xyz', 'test', 2, 0, array())), array(), array(), true, false)
182	Invalid argument supplied for foreach()
183	
184	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
185	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
186	
187	9) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #18 ('test', true, true, array('abc', 'xyz'), array(array('abc', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...), Mock_IUser_6288abb4 Object (...))), array('xyz', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...), Mock_IUser_6288abb4 Object (...)))), array(), array(array('Test One', array(0, 'test1')), array('Test Two', array(0, 'test2'))), false, false)
188	Invalid argument supplied for foreach()
189	
190	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
191	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
192	
193	10) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #19 ('test', true, false, array('abc', 'xyz'), array(array('abc', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...), Mock_IUser_6288abb4 Object (...))), array('xyz', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...), Mock_IUser_6288abb4 Object (...)))), array(), array(), true, false)
194	Invalid argument supplied for foreach()
195	
196	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
197	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
198	
199	11) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #20 ('test', true, true, array('abc', 'xyz'), array(array('abc', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...))), array('xyz', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...)))), array(array('Test One', array(0, 'test'))), array(array('Test Two', array(0, 'test2'))), false, false)
200	Invalid argument supplied for foreach()
201	
202	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
203	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
204	
205	12) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #21 ('test', true, false, array('abc', 'xyz'), array(array('abc', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...))), array('xyz', 'test', 2, 0, array(Mock_IUser_6288abb4 Object (...)))), array(array('Test One', array(0, 'test'))), array(), true, false)
206	Invalid argument supplied for foreach()
207	
208	/drone/src/lib/private/Collaboration/Collaborators/UserPlugin.php:83
209	/drone/src/tests/lib/Collaboration/Collaborators/UserPluginTest.php:461
```

**Comment by rullzer** [comment]
Ok it seems I just don't get the tests...
@MorrisJobke @nickvergessen could one of you have a look and try?

**Comment by MorrisJobke** [comment]
Underlying code was changed with #21452

**Comment by MorrisJobke** [comment]
I merged the changes from this PR with the enhancements from #21452 to also filter out disabled users properly.

Ready for review.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
What?

```
/github/workspace/apps/user_status/lib/Db/UserStatusMapper.php:84:41:error - ImplicitToStringCast: Argument 2 of OCP\DB\QueryBuilder\IExpressionBuilder::notIn expects array<array-key, mixed>|string, OCP\DB\QueryBuilder\IParameter provided with a __toString method
```

**Comment by MorrisJobke** [comment]
Seems that this came in via another PR.

**Comment by MorrisJobke** [comment]
> Seems that this came in via another PR.

It isn't on master :/

**Comment by MorrisJobke** [comment]
@kesselb Any idea?

**Comment by kesselb** [comment]
@MorrisJobke similar to https://github.com/nextcloud/server/pull/22257. But warning is unrelated to this pr.

**Pull request merged**